File: ios/Classes/StarMicronicsPrinterPlugin.swift
import Flutter
import UIKit
import StarIO10

public class StarMicronicsPrinterPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "star_micronics_printer", binaryMessenger: registrar.messenger())
        let instance = StarMicronicsPrinterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "searchPrinters":
            searchPrinters(call: call, result: result)
        case "getStatus":
            getStatus(call: call, result: result)
        case "print":
            print(call: call, result: result)
        case "printCommands":
            printCommands(call: call, result: result)
        case "openCashDrawer":
            openCashDrawer(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Search Printers

    private func searchPrinters(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let timeout = args["timeout"] as? Int else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Invalid timeout", details: nil))
            return
        }

        var printers: [[String: Any]] = []
        let manager = StarDeviceDiscoveryManagerFactory.create(interfaceTypes: [.lan, .bluetooth, .usb])

        manager.discoveryTime = TimeInterval(timeout) / 1000.0

        manager.callback = { (printer: StarPrinter) in
            let printerInfo: [String: Any] = [
                "model": printer.information?.model ?? "Unknown",
                "identifier": printer.connectionSettings.identifier,
                "interfaceType": self.interfaceTypeToString(printer.connectionSettings.interfaceType),
                "emulation": printer.information?.emulation.rawValue ?? ""
            ]
            printers.append(printerInfo)
        }

        manager.startDiscovery { error in
            if let error = error {
                result(FlutterError(code: "SEARCH_ERROR", message: error.localizedDescription, details: nil))
            } else {
                result(printers)
            }
        }
    }

    // MARK: - Get Status

    private func getStatus(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let settings = createConnectionSettings(from: args) else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Invalid connection settings", details: nil))
            return
        }

        let printer = StarPrinter(settings)

        Task {
            do {
                try await printer.open()
                let status = try await printer.getStatus()
                try await printer.close()

                let statusMap: [String: Any] = [
                    "online": !(status.coverOpen == .open || status.paper == .empty),
                    "coverOpen": status.coverOpen == .open,
                    "paperEmpty": status.paper == .empty,
                    "paperNearEmpty": status.paper == .nearEmpty,
                    "drawerOpen": status.drawer == .open
                ]

                await MainActor.run {
                    result(statusMap)
                }
            } catch {
                await MainActor.run {
                    result(FlutterError(code: "STATUS_ERROR", message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    // MARK: - Print

    private func print(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let settings = createConnectionSettings(from: args),
              let commandString = args["command"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Invalid arguments", details: nil))
            return
        }

        let printer = StarPrinter(settings)

        Task {
            do {
                try await printer.open()

                let builder = StarXpandCommand.StarXpandCommandBuilder()
                builder.addDocument(StarXpandCommand.DocumentBuilder()
                    .addPrinter(StarXpandCommand.PrinterBuilder()
                        .actionPrintText(commandString)
                        .actionCut(.partial)
                    )
                )

                let command = builder.getCommands()
                try await printer.print(command: command)
                try await printer.close()

                await MainActor.run {
                    result(true)
                }
            } catch {
                await MainActor.run {
                    result(FlutterError(code: "PRINT_ERROR", message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    // MARK: - Print Commands

    private func printCommands(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let settings = createConnectionSettings(from: args),
              let commands = args["commands"] as? [[String: Any]] else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Invalid arguments", details: nil))
            return
        }

        let printer = StarPrinter(settings)

        Task {
            do {
                try await printer.open()

                let builder = StarXpandCommand.StarXpandCommandBuilder()
                let printerBuilder = StarXpandCommand.PrinterBuilder()

                processCommands(printerBuilder: printerBuilder, commands: commands)

                builder.addDocument(StarXpandCommand.DocumentBuilder().addPrinter(printerBuilder))
                let command = builder.getCommands()

                try await printer.print(command: command)
                try await printer.close()

                await MainActor.run {
                    result(true)
                }
            } catch {
                await MainActor.run {
                    result(FlutterError(code: "PRINT_COMMANDS_ERROR", message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    // MARK: - Process Commands

    private func processCommands(printerBuilder: StarXpandCommand.PrinterBuilder, commands: [[String: Any]]) {
        for cmd in commands {
            // Text commands
            if let text = cmd["appendText"] as? String {
                printerBuilder.actionPrintText(text)
            }
            else if let text = cmd["appendTextMagnified"] as? String {
                let width = cmd["width"] as? Int ?? 1
                let height = cmd["height"] as? Int ?? 1
                printerBuilder
                    .styleMagnification(StarXpandCommand.MagnificationParameter(width: width, height: height))
                    .actionPrintText(text)
                    .styleMagnification(StarXpandCommand.MagnificationParameter(width: 1, height: 1))
            }
            else if let text = cmd["appendTextBold"] as? String {
                printerBuilder
                    .styleBold(true)
                    .actionPrintText(text)
                    .styleBold(false)
            }
            else if let text = cmd["appendTextUnderline"] as? String {
                printerBuilder
                    .styleUnderLine(true)
                    .actionPrintText(text)
                    .styleUnderLine(false)
            }
            else if let text = cmd["appendTextInvert"] as? String {
                printerBuilder
                    .styleInvert(true)
                    .actionPrintText(text)
                    .styleInvert(false)
            }

            // Alignment
            else if let alignment = cmd["setAlignment"] as? String {
                printerBuilder.styleAlignment(parseAlignment(alignment))
            }

            // Styles
            else if let bold = cmd["setBold"] as? Bool {
                printerBuilder.styleBold(bold)
            }
            else if let underline = cmd["setUnderline"] as? Bool {
                printerBuilder.styleUnderLine(underline)
            }
            else if let invert = cmd["setInvert"] as? Bool {
                printerBuilder.styleInvert(invert)
            }
            else if cmd["setMagnification"] != nil {
                let width = cmd["width"] as? Int ?? 1
                let height = cmd["height"] as? Int ?? 1
                printerBuilder.styleMagnification(StarXpandCommand.MagnificationParameter(width: width, height: height))
            }
            else if cmd["resetStyles"] != nil {
                printerBuilder
                    .styleBold(false)
                    .styleUnderLine(false)
                    .styleInvert(false)
                    .styleMagnification(StarXpandCommand.MagnificationParameter(width: 1, height: 1))
            }

            // Paper control
            else if let cutType = cmd["appendCutPaper"] as? String {
                printerBuilder.actionCut(parseCutType(cutType))
            }
            else if let lines = cmd["feedLine"] as? Int {
                printerBuilder.actionFeedLine(lines)
            }
            else if let units = cmd["feedUnit"] as? Int {
                printerBuilder.actionFeed(Double(units))
            }

            // Barcode
            else if let data = cmd["appendBarcode"] as? String {
                let symbology = parseBarcodeSymbology(cmd["symbology"] as? String)
                let height = cmd["height"] as? Double ?? 40.0
                let hri = cmd["hri"] as? Bool ?? false

                let parameter = StarXpandCommand.Printer.BarcodeParameter(content: data, symbology: symbology)
                    .setHeight(height)
                    .setPrintHRI(hri)

                printerBuilder.actionPrintBarcode(parameter)
            }

            // QR Code
            else if let data = cmd["appendQrCode"] as? String {
                let level = parseQrCodeLevel(cmd["level"] as? String)
                let cellSize = cmd["cellSize"] as? Int ?? 8

                let parameter = StarXpandCommand.Printer.QRCodeParameter(content: data)
                    .setLevel(level)
                    .setCellSize(cellSize)

                printerBuilder.actionPrintQRCode(parameter)
            }

            // PDF417
            else if let data = cmd["appendPdf417"] as? String {
                let column = cmd["column"] as? Int ?? 0
                let line = cmd["line"] as? Int ?? 0
                let module = cmd["module"] as? Int ?? 2
                let aspect = cmd["aspect"] as? Int ?? 3

                let parameter = StarXpandCommand.Printer.PDF417Parameter(content: data)
                    .setColumn(column)
                    .setLine(line)
                    .setModule(module)
                    .setAspect(aspect)

                printerBuilder.actionPrintPDF417(parameter)
            }

            // Image from byte array
            else if let byteArray = cmd["appendBitmapByteArray"] as? FlutterStandardTypedData {
                if let image = UIImage(data: byteArray.data) {
                    let width = cmd["width"] as? Int ?? 576
                    let parameter = StarXpandCommand.Printer.ImageParameter(image: image, width: width)
                    printerBuilder.actionPrintImage(parameter)
                }
            }

            // Encoding
            else if let encoding = cmd["setEncoding"] as? String {
                printerBuilder.styleCharacterEncoding(parseEncoding(encoding))
            }

            // Font style
            else if let fontStyle = cmd["setFontStyle"] as? String {
                printerBuilder.styleFont(parseFontStyle(fontStyle))
            }

            // Character space
            else if let space = cmd["setCharacterSpace"] as? Double {
                printerBuilder.styleCharacterSpace(space)
            }

            // Line space
            else if let space = cmd["setLineSpace"] as? Double {
                printerBuilder.styleLineSpace(space)
            }

            // Position
            else if let position = cmd["setAbsolutePosition"] as? Double {
                printerBuilder.styleHorizontalPositionTo(position)
            }
            else if let position = cmd["setRelativePosition"] as? Double {
                printerBuilder.styleHorizontalPositionBy(position)
            }

            // Logo
            else if let logoKey = cmd["appendLogo"] as? Int {
                let size = cmd["size"] as? String ?? "normal"
                let logoParameter = StarXpandCommand.Printer.LogoParameter(keyCode: logoKey)
                printerBuilder.actionPrintLogo(logoParameter)
            }
        }
    }

    // MARK: - Open Cash Drawer

    private func openCashDrawer(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let settings = createConnectionSettings(from: args) else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Invalid connection settings", details: nil))
            return
        }

        let printer = StarPrinter(settings)

        Task {
            do {
                try await printer.open()

                let builder = StarXpandCommand.StarXpandCommandBuilder()
                builder.addDocument(StarXpandCommand.DocumentBuilder()
                    .addDrawer(StarXpandCommand.DrawerBuilder()
                        .actionOpen(StarXpandCommand.Drawer.Channel.no1)
                    )
                )

                let command = builder.getCommands()
                try await printer.print(command: command)
                try await printer.close()

                await MainActor.run {
                    result(true)
                }
            } catch {
                await MainActor.run {
                    result(FlutterError(code: "DRAWER_ERROR", message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    // MARK: - Helper Methods

    private func createConnectionSettings(from args: [String: Any]) -> StarConnectionSettings? {
        guard let interfaceTypeStr = args["interfaceType"] as? String,
              let identifier = args["identifier"] as? String else {
            return nil
        }

        let interfaceType = stringToInterfaceType(interfaceTypeStr)
        return StarConnectionSettings(interfaceType: interfaceType, identifier: identifier)
    }

    private func stringToInterfaceType(_ str: String) -> InterfaceType {
        switch str.lowercased() {
        case "lan":
            return .lan
        case "bluetooth":
            return .bluetooth
        case "bluetoothle":
            return .bluetoothLE
        case "usb":
            return .usb
        default:
            return .lan
        }
    }

    private func interfaceTypeToString(_ type: InterfaceType) -> String {
        switch type {
        case .lan:
            return "lan"
        case .bluetooth:
            return "bluetooth"
        case .bluetoothLE:
            return "bluetoothLE"
        case .usb:
            return "usb"
        @unknown default:
            return "lan"
        }
    }

    private func parseAlignment(_ alignment: String) -> StarXpandCommand.Printer.Alignment {
        switch alignment.lowercased() {
        case "center":
            return .center
        case "right":
            return .right
        default:
            return .left
        }
    }

    private func parseCutType(_ cutType: String) -> StarXpandCommand.Printer.CutType {
        switch cutType.lowercased() {
        case "fullcut":
            return .full
        case "partialcutwithfeed":
            return .partialWithFeed
        case "fullcutwithfeed":
            return .fullWithFeed
        default:
            return .partial
        }
    }

    private func parseBarcodeSymbology(_ symbology: String?) -> StarXpandCommand.Printer.BarcodeSymbology {
        guard let symbology = symbology else { return .code128 }

        switch symbology.lowercased() {
        case "code39":
            return .code39
        case "code93":
            return .code93
        case "itf":
            return .itf
        case "jan8":
            return .jan8
        case "jan13":
            return .jan13
        case "nw7":
            return .nw7
        case "upca":
            return .upcA
        case "upce":
            return .upcE
        default:
            return .code128
        }
    }

    private func parseQrCodeLevel(_ level: String?) -> StarXpandCommand.Printer.QRCodeLevel {
        guard let level = level else { return .l }

        switch level.lowercased() {
        case "h":
            return .h
        case "m":
            return .m
        case "q":
            return .q
        default:
            return .l
        }
    }

    private func parseEncoding(_ encoding: String) -> StarXpandCommand.Printer.CharacterEncodingType {
        switch encoding.uppercased() {
        case "US-ASCII":
            return .usAscii
        case "WINDOWS-1252":
            return .windows1252
        case "SHIFT-JIS":
            return .shiftJIS
        case "WINDOWS-1251":
            return .windows1251
        case "GB2312":
            return .gb2312
        case "BIG5":
            return .big5
        case "UTF-8":
            return .utf8
        default:
            return .utf8
        }
    }

    private func parseFontStyle(_ fontStyle: String) -> StarXpandCommand.Printer.FontType {
        switch fontStyle.lowercased() {
        case "a":
            return .a
        case "b":
            return .b
        default:
            return .a
        }
    }
}

File: android/src/main/kotlin/com/phonetechbd/star_micronics_printer/StarMicronicsPrinterPlugin.kt
package com.phonetechbd.star_micronics_printer

import android.content.Context
import android.graphics.BitmapFactory
import androidx.annotation.NonNull
import com.starmicronics.stario10.*
import com.starmicronics.stario10.starxpandcommand.*
import com.starmicronics.stario10.starxpandcommand.printer.Printer
import com.starmicronics.stario10.starxpandcommand.drawer.Drawer
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.*

class StarMicronicsPrinterPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "star_micronics_printer")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "searchPrinters" -> searchPrinters(call, result)
            "getStatus" -> getStatus(call, result)
            "print" -> print(call, result)
            "printCommands" -> printCommands(call, result)
            "openCashDrawer" -> openCashDrawer(call, result)
            else -> result.notImplemented()
        }
    }

    private fun searchPrinters(call: MethodCall, result: Result) {
        scope.launch {
            try {
                val timeout = call.argument<Int>("timeout") ?: 10000
                val printers = mutableListOf<Map<String, Any>>()

                val manager = StarDeviceDiscoveryManagerFactory.create(
                    listOf(InterfaceType.Lan, InterfaceType.Bluetooth, InterfaceType.Usb),
                    context
                )

                manager.discoveryTime = timeout
                manager.callback = object : StarDeviceDiscoveryManager.Callback {
                    override fun onPrinterFound(printer: StarPrinter) {
                        printers.add(
                            mapOf(
                                "model" to (printer.information?.model ?: "Unknown"),
                                "identifier" to printer.connectionSettings.identifier,
                                "interfaceType" to printer.connectionSettings.interfaceType.name.lowercase(),
                                "emulation" to (printer.information?.emulation?.name ?: "")
                            )
                        )
                    }

                    override fun onDiscoveryFinished() {
                        result.success(printers)
                    }
                }

                manager.startDiscovery()
            } catch (e: Exception) {
                result.error("SEARCH_ERROR", e.message, null)
            }
        }
    }

    private fun getStatus(call: MethodCall, result: Result) {
        scope.launch {
            try {
                val settings = createConnectionSettings(call)
                val printer = StarPrinter(settings, context)

                withContext(Dispatchers.IO) {
                    printer.openAsync().await()
                    val status = printer.getStatusAsync().await()
                    printer.closeAsync().await()

                    val statusMap = mapOf(
                        "online" to (status.coverOpen == false && status.paperEmpty == false),
                        "coverOpen" to (status.coverOpen == true),
                        "paperEmpty" to (status.paperEmpty == true),
                        "paperNearEmpty" to (status.paperNearEmpty == true),
                        "drawerOpen" to (status.drawerOpenCloseSignal == true)
                    )

                    withContext(Dispatchers.Main) {
                        result.success(statusMap)
                    }
                }
            } catch (e: Exception) {
                result.error("STATUS_ERROR", e.message, null)
            }
        }
    }

    private fun print(call: MethodCall, result: Result) {
        scope.launch {
            try {
                val settings = createConnectionSettings(call)
                val commandString = call.argument<String>("command") ?: ""
                val printer = StarPrinter(settings, context)

                withContext(Dispatchers.IO) {
                    printer.openAsync().await()
                    
                    val job = StarXpandCommandBuilder()
                    job.addDocument(
                        DocumentBuilder().addPrinter(
                            PrinterBuilder()
                                .actionPrintText(commandString)
                                .actionCut(Printer.CutType.Partial)
                        )
                    )

                    val command = job.getCommands()
                    printer.printAsync(command).await()
                    printer.closeAsync().await()

                    withContext(Dispatchers.Main) {
                        result.success(true)
                    }
                }
            } catch (e: Exception) {
                result.error("PRINT_ERROR", e.message, null)
            }
        }
    }

    private fun printCommands(call: MethodCall, result: Result) {
        scope.launch {
            try {
                val settings = createConnectionSettings(call)
                @Suppress("UNCHECKED_CAST")
                val commandsList = call.argument<List<Map<String, Any>>>("commands") ?: emptyList()
                val printer = StarPrinter(settings, context)

                withContext(Dispatchers.IO) {
                    printer.openAsync().await()
                    
                    val builder = StarXpandCommandBuilder()
                    val printerBuilder = PrinterBuilder()
                    
                    // Process commands
                    processCommands(printerBuilder, commandsList)
                    
                    builder.addDocument(DocumentBuilder().addPrinter(printerBuilder))
                    val commands = builder.getCommands()
                    
                    printer.printAsync(commands).await()
                    printer.closeAsync().await()

                    withContext(Dispatchers.Main) {
                        result.success(true)
                    }
                }
            } catch (e: Exception) {
                result.error("PRINT_COMMANDS_ERROR", e.message, null)
            }
        }
    }

    private fun processCommands(builder: PrinterBuilder, commandsList: List<Map<String, Any>>) {
        for (cmd in commandsList) {
            when {
                // Text commands
                cmd.containsKey("appendText") -> {
                    builder.actionPrintText(cmd["appendText"] as String)
                }
                cmd.containsKey("appendTextMagnified") -> {
                    val text = cmd["appendTextMagnified"] as String
                    val width = (cmd["width"] as? Int) ?: 1
                    val height = (cmd["height"] as? Int) ?: 1
                    builder.styleMagnification(MagnificationParameter(width, height))
                        .actionPrintText(text)
                        .styleMagnification(MagnificationParameter(1, 1))
                }
                cmd.containsKey("appendTextBold") -> {
                    builder.styleBold(true)
                        .actionPrintText(cmd["appendTextBold"] as String)
                        .styleBold(false)
                }
                
                // Alignment
                cmd.containsKey("setAlignment") -> {
                    val alignment = when(cmd["setAlignment"] as String) {
                        "center" -> Printer.Alignment.Center
                        "right" -> Printer.Alignment.Right
                        else -> Printer.Alignment.Left
                    }
                    builder.styleAlignment(alignment)
                }
                
                // Styles
                cmd.containsKey("setBold") -> {
                    builder.styleBold(cmd["setBold"] as Boolean)
                }
                cmd.containsKey("setMagnification") -> {
                    val width = (cmd["width"] as? Int) ?: 1
                    val height = (cmd["height"] as? Int) ?: 1
                    builder.styleMagnification(MagnificationParameter(width, height))
                }
                
                // Paper control
                cmd.containsKey("appendCutPaper") -> {
                    val cutType = when(cmd["appendCutPaper"] as String) {
                        "fullCut" -> Printer.CutType.Full
                        "partialCutWithFeed" -> Printer.CutType.PartialWithFeed
                        else -> Printer.CutType.Partial
                    }
                    builder.actionCut(cutType)
                }
                cmd.containsKey("feedLine") -> {
                    builder.actionFeedLine(cmd["feedLine"] as Int)
                }
                
                // Barcode
                cmd.containsKey("appendBarcode") -> {
                    val data = cmd["appendBarcode"] as String
                    val symbology = parseBarcodeSymbology(cmd["symbology"] as? String)
                    val height = ((cmd["height"] as? Int) ?: 40).toDouble()
                    val hri = cmd["hri"] as? Boolean ?: false
                    
                    val parameter = Printer.BarcodeParameter(data, symbology)
                        .setHeight(height)
                        .setPrintHRI(hri)
                    
                    builder.actionPrintBarcode(parameter)
                }
                
                // QR Code
                cmd.containsKey("appendQrCode") -> {
                    val data = cmd["appendQrCode"] as String
                    val level = parseQrCodeLevel(cmd["level"] as? String)
                    val cellSize = (cmd["cellSize"] as? Int) ?: 8
                    
                    val parameter = Printer.QRCodeParameter(data)
                        .setLevel(level)
                        .setCellSize(cellSize)
                    
                    builder.actionPrintQRCode(parameter)
                }
                
                // Image
                cmd.containsKey("appendBitmapByteArray") -> {
                    val byteArray = cmd["appendBitmapByteArray"] as ByteArray
                    val width = (cmd["width"] as? Int) ?: 576
                    
                    try {
                        val bitmap = BitmapFactory.decodeByteArray(byteArray, 0, byteArray.size)
                        if (bitmap != null) {
                            builder.actionPrintImage(Printer.ImageParameter(bitmap, width))
                        }
                    } catch (e: Exception) {
                        // Handle error
                    }
                }
            }
        }
    }

    private fun parseBarcodeSymbology(symbology: String?): Printer.BarcodeSymbology {
        return when(symbology?.lowercase()) {
            "code39" -> Printer.BarcodeSymbology.Code39
            "code93" -> Printer.BarcodeSymbology.Code93
            "jan8" -> Printer.BarcodeSymbology.Jan8
            "jan13" -> Printer.BarcodeSymbology.Jan13
            else -> Printer.BarcodeSymbology.Code128
        }
    }

    private fun parseQrCodeLevel(level: String?): Printer.QRCodeLevel {
        return when(level?.lowercase()) {
            "h" -> Printer.QRCodeLevel.H
            "m" -> Printer.QRCodeLevel.M
            "q" -> Printer.QRCodeLevel.Q
            else -> Printer.QRCodeLevel.L
        }
    }

    private fun openCashDrawer(call: MethodCall, result: Result) {
        scope.launch {
            try {
                val settings = createConnectionSettings(call)
                val printer = StarPrinter(settings, context)

                withContext(Dispatchers.IO) {
                    printer.openAsync().await()
                    
                    val job = StarXpandCommandBuilder()
                    job.addDocument(
                        DocumentBuilder().addDrawer(
                            DrawerBuilder().actionOpen(Drawer.Channel.No1)
                        )
                    )

                    val command = job.getCommands()
                    printer.printAsync(command).await()
                    printer.closeAsync().await()

                    withContext(Dispatchers.Main) {
                        result.success(true)
                    }
                }
            } catch (e: Exception) {
                result.error("DRAWER_ERROR", e.message, null)
            }
        }
    }

    private fun createConnectionSettings(call: MethodCall): StarConnectionSettings {
        val interfaceTypeStr = call.argument<String>("interfaceType") ?: "lan"
        val identifier = call.argument<String>("identifier") ?: ""

        val interfaceType = when (interfaceTypeStr.lowercase()) {
            "lan" -> InterfaceType.Lan
            "bluetooth" -> InterfaceType.Bluetooth
            "bluetoothle" -> InterfaceType.BluetoothLE
            "usb" -> InterfaceType.Usb
            else -> InterfaceType.Lan
        }

        return StarConnectionSettings(interfaceType, identifier)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        scope.cancel()
    }
}